### PR TITLE
Remote sectorbuilder workers

### PR DESCRIPTION
**Pull request by magik6k**

### CODE REVIEW

**Comment by whyrusleeping** [comment]
lotus_seal

File: cmd/lotus-seal-worker/main.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		log.Warnf("%+v", err)
+		return
+	}
+}
+
+var runCmd = &cli.Command{
+	Name:  "run",
+	Usage: "Start lotus worker",
+	Action: func(cctx *cli.Context) error {
+		nodeApi, closer, err := lcli.GetStorageMinerAPI(cctx)
```
Comment (whyrusleeping) [inline cmd/lotus-seal-worker/main.go]: Does this mean it can't easily be run on a separate computer?

File: cmd/lotus-seal-worker/main.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		if err != nil {
+			return err
+		}
+		if v.APIVersion != build.APIVersion {
+			return xerrors.Errorf("lotus-storage-miner API version doesn't match: local: ", api.Version{APIVersion: build.APIVersion})
+		}
+
+		go func() {
+			<-ctx.Done()
+			os.Exit(0)
```
Comment (whyrusleeping) [inline cmd/lotus-seal-worker/main.go]: calling os.Exit() here means we never execute our defers. the context cancellation should just ensure that acceptJobs terminates.

File: cmd/lotus-seal-worker/sub.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+func (w *worker) processTask(ctx context.Context, task sectorbuilder.WorkerTask) sectorbuilder.SealRes {
+	switch task.Type {
+	case sectorbuilder.WorkerPreCommit:
+	case sectorbuilder.WorkerCommit:
+	default:
+		return errRes(xerrors.Errorf("unknown task type %d", task.Type))
+	}
+
+	if err := w.fetchSector(task.SectorID, task.Type); err != nil {
```
Comment (whyrusleeping) [inline cmd/lotus-seal-worker/sub.go]: does this mean it fetches the data over http twice per sector?

File: cmd/lotus-seal-worker/main.go (hunk 1 above)
Comment (Kubuxu) [reply cmd/lotus-seal-worker/main.go]: You would probably need a fake path with just the `api` file.

File: lib/sectorbuilder/remote.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+	taskCh := make(chan WorkerTask)
+	r := &remote{
+		sealTasks: taskCh,
+		busy:      0,
+	}
+
+	sb.remoteCtr++
+	sb.remotes[sb.remoteCtr] = r
+	sb.remoteLk.Unlock()
```
Comment (whyrusleeping) [inline lib/sectorbuilder/remote.go]: probably should just defer the unlock

File: cmd/lotus-seal-worker/sub.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+	bar := pb.New64(resp.ContentLength)
+	bar.ShowPercent = true
+	bar.ShowSpeed = true
+	bar.Units = pb.U_BYTES
+
+	bar.Start()
+	defer bar.Finish()
+
+	_, err = io.Copy(out, bar.NewProxyReader(resp.Body))
```
Comment (whyrusleeping) [inline cmd/lotus-seal-worker/sub.go]: if the copy fails, we should probably delete the file

File: cmd/lotus-seal-worker/sub.go (hunk 3 above)
Comment (magik6k) [reply cmd/lotus-seal-worker/sub.go]: Yes

File: cmd/lotus-storage-miner/run.go [hunk 6]
```diff
@@ -30,7 +30,11 @@ var runCmd = &cli.Command{
 	Flags: []cli.Flag{
 		&cli.StringFlag{
 			Name:  "api",
-			Value: "",
+			Value: "2345",
+		},
+		&cli.StringFlag{
+				Name: "apihost",
+				Value: "127.0.0.1",
```
Comment (whyrusleeping) [inline cmd/lotus-storage-miner/run.go]: maybe we should just have the api flag accept a multiaddr? We can also have a fallback where if it parses as a number, use the current behavior

File: cmd/lotus-seal-worker/main.go (hunk 1 above)
Comment (magik6k) [reply cmd/lotus-seal-worker/main.go]: ^ yes, can make this nicer later

File: cmd/lotus-seal-worker/sub.go (hunk 3 above)
Comment (magik6k) [reply cmd/lotus-seal-worker/sub.go]: (can optimize later)

File: lib/sectorbuilder/remote.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+		r.busy = 0
+		r.lk.Unlock()
+	}
+}
+
+func (sb *SectorBuilder) TaskDone(ctx context.Context, task uint64, res SealRes) error {
+	sb.remoteLk.Lock()
+	rres, ok := sb.remoteResults[task]
+	if ok {
+		delete(sb.remoteResults, task)
```
Comment (whyrusleeping) [inline lib/sectorbuilder/remote.go]: maybe hold off on deleting until we get the response back? What if we call this, decide we don't want to wait around, then call it again later?

File: lib/sectorbuilder/remote.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+				delete(sb.remotes, i)
+				return
+			}
+		}
+	}()
+
+	for {
+		select {
+		case task := <-sb.sealTasks:
+			resCh := make(chan SealRes)
```
Comment (whyrusleeping) [inline lib/sectorbuilder/remote.go]: maybe buffer 1 here just in case?

**Comment by whyrusleeping** [comment]
Hrm... its not working for me, it looks like the deal provider loop that consumes sectorIncoming isnt live?

**Comment by whyrusleeping** [comment]
Ah, its failing silently (we don't check the error). Its a problem with the sector state stored on disk (that cbor input fields count thing).

File: storage/miner.go [hunk 9]
```diff
@@ -109,7 +109,11 @@ func (m *Miner) Run(ctx context.Context) error {
 	}
 
 	go fps.run(ctx)
-	go m.sectorStateLoop(ctx)
+	if err := m.sectorStateLoop(ctx); err != nil {
```
Comment (magik6k) [inline storage/miner.go]: This must be in a goroutine, otherwise you can only process one sector at a time (including sealing)

Comment (whyrusleeping) [reply storage/miner.go]: The function does some initial checks and then spins off goroutines internally. Are you sure?

Comment (magik6k) [reply storage/miner.go]: Ohwait, I thought it's a different `go something()` call

**Pull request merged**